File: server/src/main/java/io/horizen/utils/NscMethods.java
package io.horizen.utils;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.horizen.helpers.Definitions;
import io.horizen.helpers.Helper;
import io.horizen.helpers.MyGsonManager;
import org.web3j.abi.TypeDecoder;
import org.web3j.abi.datatypes.Address;
import org.web3j.abi.datatypes.generated.Bytes3;
import org.web3j.abi.datatypes.generated.Bytes32;
import org.web3j.abi.datatypes.generated.Uint32;
import org.web3j.crypto.Hash;
import org.web3j.crypto.Keys;
import org.web3j.utils.Numeric;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.regex.Pattern;

public class NscMethods {
    public static Map<String, List<String>> getNscOwnerships(String scAddress) throws Exception {
        String method;
        String abiString;
        if (scAddress == null) {
            method = "getAllKeyOwnerships()";
            // Generate the 4-byte selector
            byte[] selector = Arrays.copyOf(Hash.sha3(method.getBytes()),4);
            abiString = "0x" + Numeric.toHexStringNoPrefix(selector);
        } else {
            // Remove the "0x" prefix
            if (scAddress.startsWith("0x"))
                scAddress = scAddress.substring(2);

            if (scAddress.length() != 40 || !Pattern.matches("[0-9A-Fa-f]+", scAddress))
                throw new Exception("Invalid sc address length: {}, expected 40");

            method = "getKeyOwnerships(address)";
            byte[] selector = Arrays.copyOf(Hash.sha3(method.getBytes()),4);
            String abiMethodString = "0x" + Numeric.toHexStringNoPrefix(selector);
            abiString = abiMethodString + "000000000000000000000000" + scAddress;
        }

        String requestBody = buildNscRequestBody(abiString);
        HttpURLConnection connection = Helper.sendRequestWithAuth(Definitions.NSC_URL + "ethv1", requestBody, "user", "Horizen");

        if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
            throw new Exception(connection.getResponseCode() + " " + connection.getResponseMessage());
        }

        // Read the response body
        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();

        // Print the response body
        System.out.println("Response Body: " + response);

        // Parse the JSON response
        JsonObject responseObject = JsonParser.parseString(response.toString()).getAsJsonObject();

        // Get the result value
        String abiReturnValue = responseObject.get("result").getAsString();

        // Remove the "0x" prefix
        if (abiReturnValue.startsWith("0x")) {
            abiReturnValue = abiReturnValue.substring(2);
        }

        return getKeyOwnershipFromAbi(abiReturnValue);
    }


    public static List<String> getNscOwnerScAddresses() throws Exception {
        String method = "getKeyOwnerScAddresses()";
        byte[] selector = Arrays.copyOf(Hash.sha3(method.getBytes()),4);
        String abiString = "0x" + Numeric.toHexStringNoPrefix(selector);

        String requestBody = buildNscRequestBody(abiString);
        HttpURLConnection connection = Helper.sendRequestWithAuth(Definitions.NSC_URL + "ethv1", requestBody, "user", "Horizen");

        if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
            throw new Exception("Problem with getKeyOwnerScAddresses()");
        }

        // Read the response body
        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();

        // Print the response body
        System.out.println("Response Body: " + response);

        // Parse the JSON response
        JsonObject responseObject = JsonParser.parseString(response.toString()).getAsJsonObject();

        // Get the result value
        String abiReturnValue = responseObject.get("result").getAsString();

        // Remove the "0x" prefix
        if (abiReturnValue.startsWith("0x"))
            abiReturnValue = abiReturnValue.substring(2);

        return getOwnerScAddrFromAbi(abiReturnValue);
    }



    public static Map<String, List<String>> getKeyOwnershipFromAbi(String abiReturnValue) {
        int startDataOffset = TypeDecoder.decode(abiReturnValue.substring(0,64), Uint32.class).getValue().intValue() * 2;
        int endDataOffset = startDataOffset + 64;
        int listSize = TypeDecoder.decode(abiReturnValue.substring(startDataOffset, endDataOffset), Uint32.class).getValue().intValue();

        Map<String, List<String>> scAssociations = new HashMap<>();
        for (int i = 0; i < listSize; i++) {
            startDataOffset = endDataOffset;
            endDataOffset = startDataOffset + 64;
            Address addressPref = TypeDecoder.decode(abiReturnValue.substring(startDataOffset, endDataOffset), Address.class);
            startDataOffset = endDataOffset;
            endDataOffset = startDataOffset + 64;
            Bytes3 mca3  = TypeDecoder.decode(abiReturnValue.substring(startDataOffset, endDataOffset), Bytes3.class);
            startDataOffset = endDataOffset;
            endDataOffset = startDataOffset + 64;
            Bytes32 mca32  = TypeDecoder.decode(abiReturnValue.substring(startDataOffset, endDataOffset), Bytes32.class);

            String scAddressChecksumFmt = Keys.toChecksumAddress(addressPref.getValue());
            byte[] combinedArray = ByteBuffer.allocate(35)
                    .put(mca3.getValue())
                    .put(mca32.getValue())
                    .array();
            String mcAddr = new String(combinedArray, StandardCharsets.UTF_8);

            List<String> mcAddresses;
            if (scAssociations.containsKey(scAddressChecksumFmt)) {
                mcAddresses = scAssociations.get(scAddressChecksumFmt);
                mcAddresses.add(mcAddr);
            }
            else {
                mcAddresses = new LinkedList<>();
                mcAddresses.add(mcAddr);
                scAssociations.put(scAddressChecksumFmt, mcAddresses);
            }
        }

        return scAssociations;
    }

    public static List<String> getOwnerScAddrFromAbi(String abiReturnValue) {
        int startDataOffset = TypeDecoder.decode(abiReturnValue.substring(0,64), Uint32.class).getValue().intValue() * 2;
        int endDataOffset = startDataOffset + 64;
        int listSize = TypeDecoder.decode(abiReturnValue.substring(startDataOffset, endDataOffset), Uint32.class).getValue().intValue();

        List<String> scAddresses = new ArrayList<>();
        for (int i = 0; i < listSize; i++) {
            startDataOffset = endDataOffset;
            endDataOffset = startDataOffset + 64;
            Address addressPref = TypeDecoder.decode(abiReturnValue.substring(startDataOffset, endDataOffset), Address.class);

            String scAddressChecksumFmt = Keys.toChecksumAddress(addressPref.getValue());
            scAddresses.add(scAddressChecksumFmt);
        }

        return scAddresses;
    }

    private static String buildNscRequestBody(String abiString) {
        // Create the request body JSON object
        JsonObject requestBody = new JsonObject();

        // Set the properties of the request body
        requestBody.addProperty("jsonrpc", "2.0");
        requestBody.addProperty("method", "eth_call");

        // Create the params JSON array
        JsonArray paramsArray = new JsonArray();

        // Create the params JSON object
        JsonObject paramsObject = new JsonObject();
        paramsObject.addProperty("from", Definitions.ETH_CALL_FROM_ADDRESS);
        paramsObject.addProperty("to", "0x0000000000000000000088888888888888888888");
        paramsObject.addProperty("value", "0x00");
        paramsObject.addProperty("gasLimit", "0x21000");
        paramsObject.addProperty("maxPriorityFeePerGas", "0x900000000");
        paramsObject.addProperty("maxFeePerGas", "0x900000000");
        paramsObject.addProperty("data", abiString);

        // Add the params JSON object to the params JSON array
        paramsArray.add(paramsObject);

        // Add the params JSON array to the request body
        requestBody.add("params", paramsArray);

        // Set the id property of the request body
        requestBody.addProperty("id", 1);

        // Create Gson instance
        Gson gson = MyGsonManager.getGson();

        // Convert the Java object to JSON string
        return gson.toJson(requestBody);
    }
}

File: py/server/modules/nsc_methods.py
import json
import requests
from eth_abi import decode
from eth_utils import remove_0x_prefix, to_checksum_address, function_signature_to_4byte_selector, encode_hex
from .definitions import NSC_URL, ETH_CALL_FROM_ADDRESS
from .util_methods import print_outgoing, print_incoming, check_sc_address, hex_str_to_bytes, print_log


# not used; useful for calling http api endpoint
def get_nsc_ownerships1(sc_address=None):
    # Call eon endpoint /transaction/getKeysOwnership which call NativeSmartContract interface
    request_body = {"scAddressOpt": sc_address}
    print_outgoing("NSC", "/transaction/getKeysOwnership", request_body)
    response = requests.post(NSC_URL + "transaction/getKeysOwnership", json.dumps(request_body),
                             auth=('user', 'Horizen'))
    print_incoming("NSC", "/transaction/getKeysOwnership", response.json())
    return response.json()['result']['keysOwnership']


# invokes eth_call RPC
def get_nsc_ownerships(sc_address=None):
    if sc_address is None:
        method = 'getAllKeyOwnerships()'
        abi_str = encode_hex(function_signature_to_4byte_selector(method))
    else:
        sc_address = remove_0x_prefix(sc_address)
        check_sc_address(sc_address)
        method = 'getKeyOwnerships(address)'
        abi_method_str = encode_hex(function_signature_to_4byte_selector(method))
        abi_str = abi_method_str + "000000000000000000000000" + sc_address

    request_body = {
        "jsonrpc": "2.0",
        "method": "eth_call",
        "params": [
            {
                "from": ETH_CALL_FROM_ADDRESS,
                "to": "0x0000000000000000000088888888888888888888",
                "value": "0x00",
                "gasLimit": "0x21000",
                "maxPriorityFeePerGas": "0x900000000",
                "maxFeePerGas": "0x900000000",
                "data": abi_str
            }
        ],
        "id": 1
    }

    abi_return_value = handle_response_from_req(request_body, 'getKeysOwnership')
    return get_key_ownership_from_abi(abi_return_value)


# invokes eth_call RPC
def get_nsc_owner_sc_addresses():
    method = 'getKeyOwnerScAddresses()'
    abi_str = encode_hex(function_signature_to_4byte_selector(method))

    request_body = {
        "jsonrpc": "2.0",
        "method": "eth_call",
        "params": [
            {
                "from": ETH_CALL_FROM_ADDRESS,
                "to": "0x0000000000000000000088888888888888888888",
                "value": "0x00",
                "gasLimit": "0x21000",
                "maxPriorityFeePerGas": "0x900000000",
                "maxFeePerGas": "0x900000000",
                "data": abi_str
            }
        ],
        "id": 1
    }

    abi_return_value = handle_response_from_req(request_body, 'getKeyOwnerScAddresses')
    return get_owner_sc_addr_from_abi(abi_return_value)

def handle_response_from_req(request_body, op_tag):
    print_outgoing("NSC", "/ethv1/eth_call (" + op_tag + ")", request_body)
    response = requests.post(NSC_URL + "ethv1", json.dumps(request_body), auth=('user', 'Horizen'))
    print_incoming("NSC", "/ethv1/eth_call (" + op_tag + ")", response.json())
    abi_return_value = remove_0x_prefix(response.json()['result'])
    return abi_return_value

def get_key_ownership_from_abi(abi_return_value):
    # the location of the data part of the first (the only one in this case) parameter (dynamic type), measured in bytes
    # from the start of the return data block. In this case 32 (0x20)
    start_data_offset = decode(['uint32'], hex_str_to_bytes(abi_return_value[0:64]))[0] * 2

    end_offset = start_data_offset + 64  # read 32 bytes
    list_size = decode(['uint32'], hex_str_to_bytes(abi_return_value[start_data_offset:end_offset]))[0]

    sc_associations_dict = {}
    for i in range(list_size):
        start_offset = end_offset
        end_offset = start_offset + 192  # read (32 + 32 + 32) bytes
        (address_pref, mca3, mca32) = decode(['address', 'bytes3', 'bytes32'],
                                             hex_str_to_bytes(abi_return_value[start_offset:end_offset]))
        sc_address_checksum_fmt = to_checksum_address(address_pref)
        if sc_associations_dict.get(sc_address_checksum_fmt) is not None:
            mc_addr_list = sc_associations_dict.get(sc_address_checksum_fmt)
        else:
            sc_associations_dict[sc_address_checksum_fmt] = []
            mc_addr_list = []
        mc_addr = (mca3 + mca32).decode('utf-8')
        mc_addr_list.append(mc_addr)
        print_log("sc_addr=" + sc_address_checksum_fmt + " / mc_addr=" + mc_addr)
        sc_associations_dict[sc_address_checksum_fmt] = mc_addr_list

    # print_log(json.dumps(sc_associations_dict, indent=4))
    return sc_associations_dict


def get_owner_sc_addr_from_abi(abi_return_value):
    # the location of the data part of the first (the only one in this case) parameter (dynamic type), measured in bytes
    # from the start of the return data block. In this case 32 (0x20)
    start_data_offset = decode(['uint32'], hex_str_to_bytes(abi_return_value[0:64]))[0] * 2

    end_offset = start_data_offset + 64  # read 32 bytes
    list_size = decode(['uint32'], hex_str_to_bytes(abi_return_value[start_data_offset:end_offset]))[0]

    sc_address_list = []
    for i in range(list_size):
        start_offset = end_offset
        end_offset = start_offset + 64  # read (32) bytes
        (address_pref) = decode(['address'], hex_str_to_bytes(abi_return_value[start_offset:end_offset]))
        sc_address_checksum_fmt = to_checksum_address(address_pref[0])
        print_log("sc addr=" + sc_address_checksum_fmt)

        sc_address_list.append(sc_address_checksum_fmt)

    #print_log(json.dumps(sc_address_list))
    return sc_address_list


proposal_dict = {}
